### ISSUE TITLE
투수연봉회귀예측-어프렌티스 실습 기말과제 

### ISSUE THREAD

**Issue by BONGKUNKIM** [BONGKUNKIM/Projects#5]
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import missingno as msno
# 경고 뜨지 않게 설정
import warnings
warnings.filterwarnings('ignore')
# 그래프 설정
plt.rcParams['font.family'] = 'Malgun Gothic'
# plt.rcParams['font.family'] = 'AppleGothic'
plt.rcParams['font.size'] = 16
plt.rcParams['figure.figsize'] = 20, 10
plt.rcParams['axes.unicode_minus'] = False
# CSV 파일을 DataFrame으로 읽기
file_path = r'C:\Users\bkkim\apprentice\picher_stats_2017.csv'
df = pd.read_csv(file_path, encoding='euc-kr')
df
df.info()
#데이터의 결측치 확인
df.isna().sum()
from sklearn.model_selection import train_test_split
target_column = '연봉(2018)'
X = df.drop(target_column, axis=1)
y = df[target_column]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
# 2018년 연봉에 대한 정보확인
df["연봉(2018)"].describe()
#연봉이 높은순서로 정렬
df.sort_values(by="연봉(2018)", ascending=False).head(20)
df["연봉(2018)"].hist(bins=100)
plt.show()
plt.boxplot(df["연봉(2018)"])
plt.show()
df["연봉(2018)"].plot.bar()
plt.show()
df2 = df.drop(["선수명", "팀명"], axis=1)
df2
# 그래프 객체 생성
fig = plt.figure()

# 각 칼럼의 히스토그램을 그린다.
for i in range(len(df2.columns)) :
	
    # 서브 플롯을 생성한다.
    # 행수, 열수, 위치(1부터 시작함)
    ax = fig.add_subplot(5, 5, i+1)
      
    plt.hist(df2[df2.columns[i]])
    
    # 현재 번째의 칼럼 이름을 가져온다.
    column_name = df2.columns[i]
    
    ax.set_title(column_name)

# 그래프들 간의 간격을 최적화하여 레이아웃을 조절한다.
plt.tight_layout()
    
plt.show()
df3 = df.drop("선수명", axis=1)
df3
# 입력과 결과로 나눈다.
X = df3.drop("연봉(2018)", axis = 1)
y = df3["연봉(2018)"]

display(X)
display(y)
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import StandardScaler

encoder1 = LabelEncoder()
encoder1.fit(X["팀명"])
X["팀명"] = encoder1.transform(X["팀명"])
X["팀명"]
scaler1 = StandardScaler()
scaler1.fit(X)
X = scaler1.transform(X)
X
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor, AdaBoostRegressor, GradientBoostingRegressor
from xgboost import XGBRegressor

# 데이터의 결측치를 채우고, 특성을 변환할 파이프라인을 생성
numeric_features = X.select_dtypes(include=['int64', 'float64']).columns
categorical_features = X.select_dtypes(include=['object']).columns

numeric_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='mean')),
    ('scaler', StandardScaler())
])

categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))
])

preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numeric_features),
        ('cat', categorical_transformer, categorical_features)
    ])

# 선형 회귀 (Linear Regression)
params_linear = {}

model_linear = LinearRegression()

# 특성 전처리와 모델을 합쳐서 파이프라인 생성
linear_pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                                   ('regressor', model_linear)])

# 그리드 서치를 사용하여 최적의 하이퍼파라미터 탐색
grid_clf_linear = GridSearchCV(linear_pipeline, param_grid=params_linear, scoring="r2", cv=5)
grid_clf_linear.fit(X, y)

print("Linear Regression - 최적의 하이퍼 파라미터:", grid_clf_linear.best_params_)
print("Linear Regression - 최적의 모델 평균 성능:", grid_clf_linear.best_score_)
# 랜덤 포레스트 (Random Forest)
params_rf = {
    'regressor__n_estimators': [50, 100, 200],
    'regressor__max_depth': [None, 10, 20],
    'regressor__min_samples_split': [2, 5, 10],
    'regressor__min_samples_leaf': [1, 2, 4]
}

model_rf = RandomForestRegressor()
rf_pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                                ('regressor', model_rf)])

grid_clf_rf = GridSearchCV(rf_pipeline, param_grid=params_rf, scoring="r2", cv=5)
grid_clf_rf.fit(X, y)

print("Random Forest - 최적의 하이퍼 파라미터:", grid_clf_rf.best_params_)
print("Random Forest - 최적의 모델 평균 성능:", grid_clf_rf.best_score_)
# 그래디언트 부스팅 (Gradient Boosting)
params_gb = {
    'regressor__n_estimators': [50, 100, 200],
    'regressor__learning_rate': [0.01, 0.1, 1],
    'regressor__max_depth': [3, 5, 7]
}

model_gb = GradientBoostingRegressor()
gb_pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                                ('regressor', model_gb)])

grid_clf_gb = GridSearchCV(gb_pipeline, param_grid=params_gb, scoring="r2", cv=5)
grid_clf_gb.fit(X, y)

print("Gradient Boosting - 최적의 하이퍼 파라미터:", grid_clf_gb.best_params_)
print("Gradient Boosting - 최적의 모델 평균 성능:", grid_clf_gb.best_score_)
# 에이다부스트 (AdaBoost)
params_ada = {
    'regressor__n_estimators': [50, 100, 200],
    'regressor__learning_rate': [0.01, 0.1, 1]
}

model_ada = AdaBoostRegressor()
ada_pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                                 ('regressor', model_ada)])

grid_clf_ada = GridSearchCV(ada_pipeline, param_grid=params_ada, scoring="r2", cv=5)
grid_clf_ada.fit(X, y)

print("AdaBoost - 최적의 하이퍼 파라미터:", grid_clf_ada.best_params_)
print("AdaBoost - 최적의 모델 평균 성능:", grid_clf_ada.best_score_)
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.model_selection import train_test_split

# 데이터를 훈련 세트와 테스트 세트로 나누기
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 선형 회귀 모델 평가
linear_pred = grid_clf_linear.predict(X_test)
linear_r2 = r2_score(y_test, linear_pred)
linear_mse = mean_squared_error(y_test, linear_pred)

print("Linear Regression - R^2 Score:", linear_r2)
print("Linear Regression - Mean Squared Error:", linear_mse)

# 랜덤 포레스트 모델 평가
rf_pred = grid_clf_rf.predict(X_test)
rf_r2 = r2_score(y_test, rf_pred)
rf_mse = mean_squared_error(y_test, rf_pred)

print("Random Forest - R^2 Score:", rf_r2)
print("Random Forest - Mean Squared Error:", rf_mse)

# 그래디언트 부스팅 모델 평가
gb_pred = grid_clf_gb.predict(X_test)
gb_r2 = r2_score(y_test, gb_pred)
gb_mse = mean_squared_error(y_test, gb_pred)

print("Gradient Boosting - R^2 Score:", gb_r2)
print("Gradient Boosting - Mean Squared Error:", gb_mse)

# 에이다부스트 모델 평가
ada_pred = grid_clf_ada.predict(X_test)
ada_r2 = r2_score(y_test, ada_pred)
ada_mse = mean_squared_error(y_test, ada_pred)

print("AdaBoost - R^2 Score:", ada_r2)
print("AdaBoost - Mean Squared Error:", ada_mse)
# 튜닝된 Gradient Boosting 모델
params_gb = {
    'regressor__n_estimators': [200],
    'regressor__learning_rate': [0.01],
    'regressor__max_depth': [5]
}

model_gb = GradientBoostingRegressor()
gb_pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                                ('regressor', model_gb)])

grid_clf_gb = GridSearchCV(gb_pipeline, param_grid=params_gb, scoring="r2", cv=5)
grid_clf_gb.fit(X, y)

# 튜닝된 Gradient Boosting 모델 평가
gb_pred = grid_clf_gb.predict(X_test)
gb_r2 = r2_score(y_test, gb_pred)
gb_mse = mean_squared_error(y_test, gb_pred)

print("Gradient Boosting - R^2 Score:", gb_r2)
print("Gradient Boosting - Mean Squared Error:", gb_mse)
# 튜닝된 Gradient Boosting 모델
params_gb = {
    'regressor__n_estimators': [100, 200, 300, 400, 500],
    'regressor__learning_rate': [0.01, 0.001, 0.0001],
    'regressor__max_depth': [3, 5, 7, 10]
}

model_gb = GradientBoostingRegressor()
gb_pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                                ('regressor', model_gb)])

grid_clf_gb = GridSearchCV(gb_pipeline, param_grid=params_gb, scoring="r2", cv=5)
grid_clf_gb.fit(X, y)

# 튜닝된 Gradient Boosting 모델 평가
gb_pred = grid_clf_gb.predict(X_test)
gb_r2 = r2_score(y_test, gb_pred)
gb_mse = mean_squared_error(y_test, gb_pred)

print("Gradient Boosting - R^2 Score:", gb_r2)
print("Gradient Boosting - Mean Squared Error:", gb_mse)
